ipn/local.go

```diff
@@ -391,26 +391,35 @@ func (b *LocalBackend) Start(opts Options) error {
 // updateFilter updates the packet filter in wgengine based on the
 // given netMap and user preferences.
 func (b *LocalBackend) updateFilter(netMap *controlclient.NetworkMap) {
-	// TODO(apenwarr): don't replace filter at all if unchanged.
-	// TODO(apenwarr): print a diff instead of full filter.
 	if netMap == nil {
 		// Not configured yet, block everything
 		b.logf("netmap packet filter: (not ready yet)")
 		b.e.SetFilter(filter.NewAllowNone(b.logf))
-	} else if b.shieldsAreUp() {
+		return
+	}
+
+	b.mu.Lock()
+	advRoutes := b.prefs.AdvertiseRoutes
+	b.mu.Unlock()
+	localNets := wgCIDRsToFilter(netMap.Addresses, advRoutes)
+
+	if b.shieldsAreUp() {
 		// Shields up, block everything
 		b.logf("netmap packet filter: (shields up)")
-		b.e.SetFilter(filter.NewAllowNone(b.logf))
+		b.e.SetFilter(filter.New(nil, localNets, nil, b.logf))
```

Comment:
whenever I see a call site like this, I start rethinking APIs nil false 0 0 true nil

Reply:
At minimum, put the nils in local variables so it's obvious what is nil.

Or `nil /* foo */, localNets, nil /* bar */` but that looks so foreign in Go.

Reply:
Agree, this API is getting a little unwieldy. For now I'll define local vars to make the call site easier to read.